### PR TITLE
Fix check for subpath source

### LINKED ISSUE #105472: subpath sources never match when containers are re-created
When we merged - https://github.com/kubernetes/kubernetes/pull/89629/files#diff-22883616376d6c596f9b6338dc1e7bc72064315dfe8d4ebc784946ba58e1c0a1R117 we introduced comparison of subpath sources:

```golang
if mntInfo.Root != subpath.Path {
  // unmount first
}
```

But subpaths source comparison never works properly for all volume types, even if subpath and source remains unchaged. This is because we are comparing wrong things. 

For a given volume - `subpath.Path` is usually:

```
/var/lib/kubelet/pods/417178d6-f74c-4cab-b49d-73c1c1052777/volumes/kubernetes.io~csi/pvc-04501ad5-1295-4430-b672-5184045d5d24/mount/foobar
```

But according `/proc/self/mountinfo` - the root is - `/foobar`. 

This means that - if creation of a container times out, then another container can not start until  old container exits, because subapth comparison above will be false and kubelet will try to umount the subpath bind mount and it will keep failing with "device is busy" errors. So subpath bind mount can not be reused at all. 

This is sort of expensive, for containers that use selinux. Because it means each container creation will potentially run recursive chcon again on the entire subpath etc.

### PR DESCRIPTION
**Pull request by gnufied**
Fix https://github.com/kubernetes/kubernetes/issues/105472

I have verified that this fixes CSI/NFS and other volume types and at the same time it keeps older behaviour for configmaps/secrets etc. This code relies on the fact that most volume types create a mount point in `/var/lib/kubelet/pods/xxxx/volumes/` when being mounted. The check here won't work for volume types that don't create mount point for main volume (but it is not a regression, at least). For configmaps etc - no bind mount is created unless user specifies subpaths.

For NFS volume with fix:

```
9660:I1006 18:38:53.488871  766496 subpath_linux.go:122] Skipping bind-mounting subpath /var/lib/kubelet/pods/f7b2da79-86eb-4a66-a9f8-5f3853c708be/volume-subpaths/manual-pv/sandbox/0, volumePath: /var/lib/kubelet/pods/f7b2da79-86eb-4a66-a9f8-5f3853c708be/volumes/kubernetes.io~nfs/manual-pv, path: /var/lib/kubelet/pods/f7b2da79-86eb-4a66-a9f8-5f3853c708be/volumes/kubernetes.io~nfs/manual-pv/foobar
```

```release-note
Do not unmount and mount subpath bind mounts during container creation unless bind mount changes
```

### CODE REVIEW

File: pkg/volume/util/hostutil/hostutil_linux.go [hunk 1]
```diff
@@ -164,6 +164,28 @@ func (hu *HostUtil) FindMountInfo(path string) (mount.MountInfo, error) {
...
+		return mount.MountInfo{}, err
+	}
+
+	// process /proc/xxx/mountinfo in backward order and find the first mount
+	// point that is prefix of 'path' - that's the mount where path resides
+	var info *mount.MountInfo
+	for i := len(infos) - 1; i >= 0; i-- {
+		if path == infos[i].MountPoint {
+			info = &infos[i]
+			break
```
Comment (bertinatto) [inline pkg/volume/util/hostutil/hostutil_linux.go:180]: nit: as a shorter alternative, we could return here and remove the if clause below

File: pkg/volume/util/subpath/subpath_linux.go [hunk 2]
```diff
@@ -116,6 +116,13 @@ func prepareSubpathTarget(mounter mount.Interface, subpath Subpath) (bool, strin
 			return false, "", fmt.Errorf("error calling findMountInfo for %s: %s", bindPathTarget, err)
 		}
 		if mntInfo.Root != subpath.Path {
+			volumeMountInfo, err := linuxHostUtil.FindExactMountInfo(subpath.VolumePath)
+			if err == nil && mount.PathWithinBase(subpath.Path, subpath.VolumePath) &&
+				(volumeMountInfo.Major == mntInfo.Major && volumeMountInfo.Minor == mntInfo.Minor) {
+				klog.V(5).Infof("Skipping bind-mounting subpath %s, volumePath: %s, path: %s", bindPathTarget, subpath.VolumePath, subpath.Path)
+				return true, bindPathTarget, nil
+			}
```
Comment (bertinatto) [inline pkg/volume/util/subpath/subpath_linux.go:124]: Here the error is ignored because of ConfigMaps?

**Review by msau42 — COMMENTED** [review]
Is it possible to add an e2e test where it would have failed for persistent volume types without this fix?

**Comment by msau42** [comment]
> Is it possible to add an e2e test where it would have failed for persistent volume types without this fix?

Do we need a test similar to https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/subpath.go#L320?

**Comment by gnufied** [comment]
@msau42 @jsafrane I have updated the code to use `os.SameFile` to compare bindmount target (i.e subpath bind mount kubelet creates) with subpath directory/file inside the volume. Since a bind mount has same inode number, device numbers as original file - this check should work. If subpath inside volume however changes (but bind mount does not), then they will mismatch and whenever container is restarted, it will cause old subpath bind mount to unmount and mounted again. PTAL.

File: pkg/volume/util/subpath/subpath_linux_test.go [hunk 3]
```diff
@@ -921,6 +921,144 @@ func TestBindSubPath(t *testing.T) {
...
+				}
+				return mounts, volpath, subpath, nil
+			},
+			expectError:  false,
+			expectAction: []mount.FakeAction{},
+			mountExists:  true,
+		},
+	}
+	for _, test := range tests {
+		klog.V(4).Infof("test %q", test.name)
```
Comment (jsafrane) [inline pkg/volume/util/subpath/subpath_linux_test.go:986]: Just a suggestion for temp dirs + defer for `os.RemoveAll`
```
t.Run(test.name, func (t *testing.T){
    base, err = ioutil.TempDir()
   if err {...}
    defer os.RemoveAll(base)
```

Comment (gnufied) [reply pkg/volume/util/subpath/subpath_linux_test.go:986]: fixed.

**Comment by jsafrane** [comment]
lgtm. Giving @msau42 a chance to review.

**Comment by msau42** [comment]
/lgtm
/hold
can you update the release note?

**Comment by gnufied** [comment]
done.
/hold cancel

**Comment by gnufied** [comment]
Unrelated failure:

```
goroutine 676 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic({0x50c15a0, 0x89ad100})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xe6
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x461107})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0xb0
panic({0x50c15a0, 0x89ad100})
	/usr/local/go/src/runtime/panic.go:1047 +0x266
k8s.io/kubernetes/pkg/kubelet.TestSyncLabels.func1.2()
```
/retest

**Pull request merged**